File: pt-3/95.unique-binary-search-trees-ii.cpp
#include <algorithm>
#include <cctype>
#include <climits>
#include <cmath>
#include <cstdlib>
#include <functional>
#include <numeric>
#include <queue>
#include <sstream>
#include <string>
#include <unordered_map>
#include <utility>
#include <vector>
using namespace std;

// Definition for a binary tree node.
// struct TreeNode {
//     int val;
//     TreeNode *left;
//     TreeNode *right;
//     TreeNode() : val(0), left(nullptr), right(nullptr) {}
//     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
//     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
// };


class Solution {
public:
    vector<TreeNode*> generateTrees(int n) {
        /*
        Approach:
        For each root value, recursively generate every valid left subtree from
        smaller values and every valid right subtree from larger values, then
        combine each pair under a new root. Empty subtree is represented by a
        single nullptr option so leaf combinations work naturally.

        C++ notes:
        unordered_map<string, vector<TreeNode*>> memoizes subtree ranges similar
        to Python lru_cache. The returned subtrees may share cached child
        pointers, which matches the source algorithm and is fine for LeetCode's
        read-only validation.

        Complexity: O(C_n * n) time and space for Catalan number C_n trees.
        */
        unordered_map<string, vector<TreeNode*>> memo;
        function<vector<TreeNode*>(int, int)> build = [&](int lo, int hi) -> vector<TreeNode*> {
            if (lo > hi) return {nullptr};
            string key = to_string(lo) + "," + to_string(hi);
            if (memo.count(key)) return memo[key];
            vector<TreeNode*> trees;
            for (int rootVal = lo; rootVal <= hi; ++rootVal) {
                vector<TreeNode*> leftTrees = build(lo, rootVal - 1);
                vector<TreeNode*> rightTrees = build(rootVal + 1, hi);
                for (TreeNode* left : leftTrees) {
                    for (TreeNode* right : rightTrees) {
                        TreeNode* root = new TreeNode(rootVal);
                        root->left = left;
                        root->right = right;
                        trees.push_back(root);
                    }
                }
            }
            memo[key] = trees;
            return trees;
        };
        return build(1, n);
    }
};

File: Meta/95.unique-binary-search-trees-ii.go
package leetcode

// GenerateTrees95 recursively chooses each value as root, combines every left
// subtree with every right subtree, and memoizes ranges. Empty subtrees are
// represented by a single nil option so leaf combinations work.
//
// Time/space: O(Cn*n) for Cn generated trees.
func GenerateTrees95(n int) []*TreeNode {
	type key struct{ lo, hi int }
	memo := map[key][]*TreeNode{}
	var build func(int, int) []*TreeNode
	build = func(lo, hi int) []*TreeNode {
		if lo > hi {
			return []*TreeNode{nil}
		}
		k := key{lo, hi}
		if v, ok := memo[k]; ok {
			return v
		}
		trees := []*TreeNode{}
		for rootVal := lo; rootVal <= hi; rootVal++ {
			for _, left := range build(lo, rootVal-1) {
				for _, right := range build(rootVal+1, hi) {
					root := &TreeNode{Val: rootVal, Left: left, Right: right}
					trees = append(trees, root)
				}
			}
		}
		memo[k] = trees
		return trees
	}
	return build(1, n)
}
